### PR TITLE
Pool Actions

### PR DESCRIPTION
**Pull request by Bromel777**

### CODE REVIEW

File: src/Cardano/PoolActions.hs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import           Plutus.V1.Ledger.Value
+import qualified PlutusTx.AssocMap   as Map
+import           Ledger.Scripts
+import qualified Ledger.Typed.Scripts.Validators as Validators
+import qualified PlutusTx
+import           Data.List
+import           Data.Maybe
+
+createPool :: PoolParams -> FullTxOut -> FullTxOut -> FullTxOut -> TxCandidate
+createPool pp@PoolParams{..} inputWithX inputWithY inputWithNft = let
```
Comment (GusevTimofey) [inline src/Cardano/PoolActions.hs:17]: `let` should start from a new line

File: src/Cardano/PoolActions.hs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  commonValue   = foldl (\acc curOutput -> acc <> (fullTxOutValue curOutput)) (Value Map.empty) [inputWithX, inputWithY, inputWithNft]
+  toInputs      = map fromPkTxOut2TxIn [inputWithX, inputWithY, inputWithNft]
+  newPoolOutput = TxOutCandidate {
+		txOutCandidateAddress = Validators.validatorAddress poolInstance,
```
Comment (GusevTimofey) [inline src/Cardano/PoolActions.hs:21]: `commas` should start from a new line

File: src/Cardano/PoolActions.hs (hunk 1 above)
Comment (oskin1) [reply src/Cardano/PoolActions.hs:17]: What if X, Y are in a single utxo or even X, Y and NFT? Just pass a list of inputs and check that all the required assets are available.

Comment (oskin1) [reply src/Cardano/PoolActions.hs:17]: Also, what about error cases? Should `createPool` always succeed?

File: src/Cardano/PoolActions.hs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		txOutCandidateValue = commonValue,
+		txOutCandidateDatum = Just $ Datum (PlutusTx.toBuiltinData pp)
+  } in TxCandidate {
```
Comment (oskin1) [inline src/Cardano/PoolActions.hs:24]: Better use `.. where ..` instead of `let .. in ..`

File: src/Cardano/PoolActions.hs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		txCandidateInputs = toInputs,
+		txCandidateOutputs = [newPoolOutput]
+  }
+
+initPool :: PoolParams -> PoolParams -> FullTxOut -> TxCandidate
```
Comment (oskin1) [inline src/Cardano/PoolActions.hs:29]: Pool params are always the same for a given pool, why old and new? Also, error cases?

File: src/Cardano/Utils.hs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module Cardano.Utils where
+
+import            Plutus.V1.Ledger.Tx
+import            Ledger
+import            Ledger.Tx
+import            Cardano.Models
+
+fromPkTxOut2TxIn :: FullTxOut -> FullTxIn
+fromPkTxOut2TxIn out@FullTxOut{..} = FullTxIn {
```
Comment (oskin1) [inline src/Cardano/Utils.hs:9]: Why do we need these?

**Review by oskin1 — CHANGES_REQUESTED** [review]
And take case of the formatting.

File: src/Cardano/PoolActions.hs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import           Data.List
+import           Data.Maybe
+
+newtype TxCandidateCreationErr = TxCandidateCreationErr { errMsg :: String }
+
+coinNotFoundError :: Coin a -> TxCandidateCreationErr
+coinNotFoundError coin =
+  TxCandidateCreationErr ( "Coin " ++ show coin ++ " not presented" )
+
+createPool :: PoolParams -> [(FullTxOut, Maybe Redeemer)] -> Address -> Either TxCandidateCreationErr TxCandidate
```
Comment (oskin1) [inline src/Cardano/PoolActions.hs:38]: Why not using FullTxIn here? `cast2Input` is not the responsibility of `createPool`.

File: src/Cardano/PoolActions.hs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  coinY   <-
+    maybeToRight (coinNotFoundError poolY) (getAssetAmountFromValueByCoin poolY commonValue)
+
+  coinNft <-
+    maybeToRight (coinNotFoundError poolNft) (getAssetAmountFromValueByCoin poolNft commonValue)
+
+  let
+    outputPoolValue =
+      coinXValue <> coinYValue <> coinNftValue
+        where
```
Comment (oskin1) [inline src/Cardano/PoolActions.hs:63]: Wrong formatting.

File: src/Cardano/PoolActions.hs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+          coinXValue = assetAmountValue coinX
+          coinYValue = assetAmountValue coinY
+          coinNftValue = assetAmountValue coinNft
+
+    outputWithPool =
+      TxOutCandidate
+        { txOutCandidateAddress = Validators.validatorAddress poolInstance
+        , txOutCandidateValue = outputPoolValue
```
Comment (oskin1) [inline src/Cardano/PoolActions.hs:71]: Please align `=` vertically.

**Pull request closed without merging**